### PR TITLE
Add wildcard_match_arm lint

### PR DESCRIPTION
**Pull request by Aehmlo**
This lint prevents using a wildcard in a match arm. Implemented as a restriction currently, because this is pretty much an edge case. See #3649 for more information.

Didn't add any tests because I wasn't sure how, but if someone wants to point me in the right direction, I'd be happy to!

### CODE REVIEW

File: clippy_lints/src/matches.rs [hunk 1]
```diff
@@ -438,6 +459,22 @@ fn check_wild_err_arm(cx: &LateContext<'_, '_>, ex: &Expr, arms: &[Arm]) {
...
 
+fn check_wild_arm(cx: &LateContext<'_, '_>, ex: &Expr, arms: &[Arm]) {
+    let ex_ty = walk_ptrs_ty(cx.tables.expr_ty(ex));
+    if match_type(cx, ex_ty, &paths::RESULT) {
+        for arm in arms {
+            if is_wild(&arm.pats[0]) {
+                span_note_and_lint(cx,
+                    MATCH_WILD,
+                    arm.pats[0].span,
+                    "Wildcard match will miss any future added variants.",
```
Comment (flip1995) [inline clippy_lints/src/matches.rs]: NIT: 
```suggestion
                    "wildcard match will miss any future added variants",
```

File: clippy_lints/src/matches.rs [hunk 2]
```diff
@@ -187,6 +187,25 @@ declare_clippy_lint! {
...
+///
+/// **Example:**
+/// ```rust
+/// match x {
+///     A => {},
+///     _ => {}
+/// }
+/// ```
+declare_clippy_lint! {
+    pub MATCH_WILD,
```
Comment (flip1995) [inline clippy_lints/src/matches.rs]:
```suggestion
    pub WILDCARD_MATCH_ARM,
```
lint names should build a sentence (more or less) if they're put after an "allow": #2845 
"allow wildcard match arm" is more a sentence than "allow match wild".

**Comment by Aehmlo** [comment]
Update: I've made some changes (including the above), but it'll be a few days until I can finish implementing the tests. I'll push everything once that's done.

**Comment by Aehmlo** [comment]
Wrong button! I think this should be good to go now.

**Comment by flip1995** [comment]
Thanks! 

Only a `rustfmt` run is missing, then this will be good to go.
https://travis-ci.com/rust-lang/rust-clippy/jobs/173071805#L1387-L1415

**Comment by Aehmlo** [comment]
Ran `rustfmt` and pushed, but it seems that there are some other nightly-related breakages that popped up.

**Comment by flip1995** [comment]
Thanks! That should be fixed by now.

@bors r+

**Comment by flip1995** [comment]
It seems that a formatting of the tests is missing.

https://travis-ci.com/rust-lang/rust-clippy/jobs/173709864#L1383-L1447

A `cargo fmt --all` inside the crate root should format everything

**Comment by Aehmlo** [comment]
Neither `cargo fmt --all` nor `cargo +nightly fmt --all` was working to format the tests, so I just ran `rustfmt` on the file directory. Regardless, it looks like the build is now passing!

**Comment by orium** [comment]
I think it would be better just to emit a lint in cases of an enum. So

```rust
match 3 {
    4 => println!(),
    _ => println!(),
};
```

would not emit a warning.

Note also that it would need to check for matching in an enum on any nesting level.  I.e.

```rust
match Some(Color::Red) {
    Some(Color::Blue) => println!(),
    Some(_) => println!(),
    None => println!(),
};
```

would emit a lint.

**Comment by Aehmlo** [comment]
I agree re: linting only on enums. I've changed the behavior and updated the lint name accordingly. I'm not familiar enough with compiler internals to implement the nested check yet, so I added a note of this limitation to the known issues. I may try to resolve this in the future if I have time, or someone else can, but I won't be able to do it in the near future.

Thanks for the insightful feedback, @orium!

**Comment by Aehmlo** [comment]
Updated lint count; should actually work next time.
